### ISSUE TITLE
cpp frontend libtorch torch::optim::SGD step memory leak win10 cu100 stable1.1

### ISSUE THREAD

**Issue by chennian32** [pytorch/pytorch#21855]
## 🐛 Bug

for (size_t epoch = 0; epoch < epoch_count; ++epoch) {
			for (auto& batch : *dataloader) {
				if (_is_exit_)
					return false;
				model_->train();				
				auto data = batch.data.to(d), targets = batch.target.to(d);	
				batch.data.reset();
				batch.target.reset();
				optimizer_->zero_grad();
				auto prediction = model_->forward(data);
				auto loss = get_loss(prediction, targets);		
				loss.backward();
				//optimizer_->step();注释掉不会泄漏，如果不注释会内存泄漏
## To Reproduce

1.
1.
1.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:
